Add terminology cheat sheet

pzehner opened the pull request:
This PR adds a new cheat sheet as suggested in #33.

Current outcome:

![Capture d’écran du 2025-03-10 14-22-58](https://github.com/user-attachments/assets/d0b0ffb7-eb84-4dfe-a75d-951af91f1ee1)

rbourgeois33 commented:
This is a great format, some remarks:

First and second line after “Kokkos,” you wrote “Default Execution space’s memory,” is that normal?

For the hierarchical structure, I think a Kokkos Team corresponds to a CUDA Block, and a Kokkos League corresponds to a CUDA Grid.
Grid of blocks = League of teams
Block of warps = Team of vectors ?
Warp of threads = Vector ? of threads

yasahi-hpc commented:
For me, it is more helpful to know which API is actually mapped to the parallelism level. For example, 

| Kokkos | OpenMPTarget  | OpenACC |
| -------- | -------- | ------- |
| `League` | `#teams distribute` |  `#loop gang` |
| `TeamThread` | `#omp simd` |  `#loop worker` |
| `ThreadVector` | `#omp simd` |  `#loop vector` |
| `TeamVector` | `#omp for simd` |  `#loop vector` |

cedricchevalier19 commented:
I do not understand the second table; where is Kokkos?

I am still not convinced by the need to talk about Cuda's `local memory` or `register.`

The last table is the most interesting part, but we should move the OpenMP target and OpenACC columns to the end and start with Cuda, HIP, and Sycl/DPC++.

pzehner commented:
> I do not understand the second table; where is Kokkos?

How would you map Kokkos on it?

> I am still not convinced by the need to talk about Cuda's local memory or register.

At least it gives equivalences of the terms between Cuda/HIP/SYCL.

cedricchevalier19 commented:
> > I do not understand the second table; where is Kokkos?
> 
> How would you map Kokkos on it?

I think one feature of the the cheat sheet is to be kept small. Here I sincerely do not understand what the table is about. I think memory and hierarchiecal parallelism are enough.

> 
> > I am still not convinced by the need to talk about Cuda's local memory or register.
> 
> At least it gives equivalences of the terms between Cuda/HIP/SYCL.

It is a Kokkos cheat sheet, these are advanced non-Kokkos terms.

In my opinion, the initial need was to make bridges between high level notions from the different programming languages. There is no one-to-one mapping on specific notions and no guarantee that Kokkos and the other languages will evolve separately.

cedricchevalier19 commented on terminology.md:
```diff
@@ -0,0 +1,48 @@
+|--------|---------------|----------------|--------|-------------|------------|
+| League | League        | --             | Grid   | Index range | ND-range   |
+| Team   | Team          | Gang           | Block  | Work group  | Work group |
+| Thread | --            | --             | Warp   | Wavefront   | Sub group  |
```
Shift OpenMP target and OpenAcc at the end.

Perhaps use "pragma" notation for directive users. (as suggested by @yasahi-hpc )

tpadioleau replied:
I agree about shifting. Actually should they also appear in the first table ?

pzehner replied:
I think I'll add them later.

cedricchevalier19 commented on terminology.md:
```diff
@@ -0,0 +1,48 @@
+|-----------------------------------------------|------------------------------|--------------------|-----------------------------|
+| `Kokkos::DefaultExecutionSpace::memory_space` | Global memory                | Global memory      | Global memory               |
+| `Kokkos::ScratchMemorySpace`                  | Shared memory                | Shared memory      | (Shared) local memory       |
+| --                                            | Local memory                 | Local memory       | --                          |
+| --                                            | Register                     | Register           | Private memory              |
```
```suggestion
```

tpadioleau replied:
I agree, as discussed, I think we can only keep lines that have Kokkos equivalent

tpadioleau commented on terminology.md:
```diff
@@ -0,0 +1,48 @@
+| Kokkos                                        | Cuda                         | HIP                | SYCL                        |
+|-----------------------------------------------|------------------------------|--------------------|-----------------------------|
+| `Kokkos::DefaultExecutionSpace::memory_space` | Global memory                | Global memory      | Global memory               |
+| `Kokkos::ScratchMemorySpace`                  | Shared memory                | Shared memory      | (Shared) local memory       |
```
It also depends on the level specified at runtime, https://kokkos.org/kokkos-core-wiki/API/core/policies/TeamPolicy.html#_CPPv4N10TeamPolicy16set_scratch_sizeERKiRKN4Impl12PerTeamValueE. I don't know if users must rely on though:
> Set the per team and per thread scratch size.
 level: set the storage level. 0 is closest cache. 1 is closest storage (e.g. high bandwidth memory)

pzehner replied:
I added a comment for this.

tpadioleau commented on terminology.md:
```diff
@@ -0,0 +1,48 @@
+| --                                            | Local memory                 | Local memory       | --                          |
+| --                                            | Register                     | Register           | Private memory              |
+| `Kokkos::SharedHostPinnedSpace`               | Pinned host memory           | Pinned host memory | Host unified shared memory  |
+| `Kokkos::SharedSpace`                         | Unified virtual memory (UVM) | Unified memory     | Unified shared memory (USM) |
```
For SYCL I think we need to specify a USM allocation of kind `shared`

tpadioleau commented on terminology.md:
```diff
@@ -0,0 +1,48 @@
+<img title="Doc" alt="Doc" src="./images/doc_txt.svg" height="25"> https://rocm.docs.amd.com/projects/HIP/en/latest/how-to/performance_guidelines.html
+<!--#endif-->
+
+## Execution
+
+| Cuda on NVIDIA                | HIP on AMD         | SYCL/OpenCL on Intel | Notes                   |
+|-------------------------------|--------------------|----------------------|-------------------------|
+| Streaming multiprocessor (SM) | Compute unit (CU)  | Compute unit         |                         |
+| Streaming processor           | Processing element | Processing element   |                         |
+| Warp                          | Wavefront          | Sub group            |                         |
+| Thread                        | Work item          | Work item            |                         |
+| NVPTX                         | AMDIL              | SPIR                 | Not strictly equivalent |
+
```
I agree, we could provide an other table about the terminology used by the different GPU vendors

tpadioleau commented on terminology.md:
```diff
@@ -0,0 +1,48 @@
+
+## Hierarchical parallelism
+
+| Kokkos | OpenMP target | OpenACC        | Cuda   | HIP         | SYCL       |
```
I think for HIP we have the same terminology as for CUDA, except warp that is indeed wavefront

tpadioleau commented on terminology.md:
```diff
@@ -0,0 +1,48 @@
+
+| Kokkos | OpenMP target | OpenACC        | Cuda   | HIP         | SYCL       |
+|--------|---------------|----------------|--------|-------------|------------|
+| League | League        | --             | Grid   | Index range | ND-range   |
```
How should the user interpret the lines ? A Kokkos league is at the same level than a CUDA grid but does not give access to all features of a CUDA grid, it is 3D for example.

pzehner replied:
I'll specify that the teams are one-dimensional.

tpadioleau reviewed (changes requested):
Should we also provide OpenMP/Threads backends ?

pzehner commented on terminology.md:
```diff
@@ -0,0 +1,48 @@
+| `Kokkos::ScratchMemorySpace`                  | Shared memory                | Shared memory      | (Shared) local memory       |
+| --                                            | Local memory                 | Local memory       | --                          |
+| --                                            | Register                     | Register           | Private memory              |
+| `Kokkos::SharedHostPinnedSpace`               | Pinned host memory           | Pinned host memory | Host unified shared memory  |
```
I remember when Damien commented the cheat sheet last June, he wanted the shared host pinned memory space to be mentioned.

pzehner commented:
Given the feedback I got, I reorganized the terminology cheat sheet to show the Kokkos mapping first. Following the idea of @yasahi-hpc, it is more API-centered.

I removed the reference to OpenMP target and OpenACC for now. There was an ambiguity in the message between "this is how Kokkos maps to these programming models," and "this is how these programming models map to Cuda/HIP/SYCL" (which is the idea I followed, hence placing them *before* Cuda/HIP/SYCL).

Then, I moved the GPU terminology equivalences part at the end, given that it is less related to Kokkos.

>>>  I do not understand the second table; where is Kokkos?
>>
>>  How would you map Kokkos on it?
>
> I think one feature of the the cheat sheet is to be kept small. Here I sincerely do not understand what the table is about. I think memory and hierarchiecal parallelism are enough.

This is is a separate cheat sheet, supposed to be less than 1 page, so as to fit in the currently empty page at the end of the printed booklet. I think it is helpful to have the GPU terminology equivalence as a Kokkos program will likely run on these GPUs. Given that the terminology is not standardized, and that it is quite hard to find these equivalences for Cuda/HIP/SYCL altogether at the same time, I think it is a good thing that a Kokkos documentation mentions it somewhere.

tpadioleau reviewed (approved):
ok for me

cedricchevalier19 commented on terminology.md:
```diff
@@ -0,0 +1,59 @@
+
+| Kokkos                                        | Cuda                         | HIP                | SYCL                               |
+|-----------------------------------------------|------------------------------|--------------------|------------------------------------|
+| `Kokkos::DefaultExecutionSpace()`             | Stream                       | Stream             | Queue                              |
```
Isn't this the default stream? The `ExecutionSpace` can be the stream or the stream on a specific device.

Perhaps add some examples on how to create streams in Kokkos can be helpful.

tpadioleau replied:
What about putting only the type `Kokkos::DefaultExecutionSpace` ?

pzehner replied:
If I'm not wrong, the type alone does not designate a stream/queue.

tpadioleau replied:
Yes true, what about `Kokkos::DefaultExecutionSpace exec_space` ? We don't specify how it was constructed

pzehner replied:
I think I can come up with something a bit more general.

tpadioleau commented on terminology.md:
```diff
@@ -0,0 +1,59 @@
+
+### Hierarchical parallelism
+
+| Kokkos                                                                         | Cuda   | HIP               | SYCL       |
+|--------------------------------------------------------------------------------|--------|-------------------|------------|
+| Instance of `Kokkos::TeamPolicy</* ... */>`                                     | Grid   | Grid, index range | ND-Range   | League
+| Instance of `Kokkos::TeamPolicy</* ... */>::member_type` (one-dimensional only) | Block  | Block, work group | Work-group | Team
+| Instance of `Kokkos::TeamThread*Range`                                          | Warp   | Warp, wavefront   | Sub-group  | SIMD chunk
+| Instance of `Kokkos::TeamVector*Range`, `Kokkos::ThreadVector*Range` | Thread | Thread, work item | Work-item  | OpenMP thread, SIMD lane
```
Shouldn't we align vertical bars ? :)

tpadioleau reviewed (approved):
Good for me
